text/RFC-0011-Add New Account Creation Mechanism on Asset Hubs.md

```diff
@@ -0,0 +1,167 @@
+does exist, the full amount of the asset can be transferred. This would mean that only the first
+asset transfer to an account has some amount debited to acquire the DOT to create the account, but
+subsequent transfers would always be in full.
+
+The main benefit of this approach is that it removes the sender's need to know about the
+desination's existence and the recipient's need to "prepare" an account by endowing it.
+
+The primary tradeoff, of course, is that transactions like "send 10 USDT" could result in fewer
+than 10 USDT arriving in the destination account. This can be solved by having the conversion be
+opt-in for the sender.
+
+Because the existential deposit is small (0.1 DOT on Asset Hub), and the user need not interact
+with the DOT in any way -- because transaction fee payment can also be handled via Asset Conversion
+-- many users may find this path convenient in avoiding transfer errors due to non-existent
+accounts or asset insufficiency.
+
+Stripping out all other asset transfer-associated logic, this RFC proposes the following logic:
+
+```rust
+fn transfer(
+    origin: OriginFor,
+    asset: AssetId,
+    destination: AccountId,
+    amount: Balance,
+    create_destination: bool,
+    ..
+) -> DispatchResult {
+    let from = ensure_signed(origin)?;
+    let details = Asset::<T, I>::get(&id).ok_or(Error::<T, I>::Unknown)?;
+    if destination.exists() || !create_destination || details.sufficient {
+        // Either the destination already exists (holds ED of DOT), the user does not want to create
+        // the destination account, or the asset class is sufficient. We can just transfer the
+        // asset as normal.
+        Self::do_transfer(asset, from, destination, amount, ..)?;
+    } else {
+        // The destination does not exist and the user has opted in to create it via a swap.
+        //
+        // We will try to swap the asset provided for the existential deposit, depositing the ED in
+        // the destination account. If the asset does not have an Asset Conversion pair with DOT or
+        // the asset amount isn't enough to acquire the existential deposit, this will fail. But we
+        // generally think (a) pairs will exist, and (b) the ED is small and UIs can easily verify
+        // that this should succeed, so failures should be rare.
+        //
+        // The swap returns the amount of the asset consumed to acquire the ED.
+        let consumed = Swap::swap_tokens_for_exact_tokens(
+            from,                // sender
+            vec![asset, dot],    // path, where `dot` is Multilocation {parents: 1, interior: Here}
+            existential_deposit, // amount_out, we need the ED for the account
+            destination,         // send_to
+            ..
+        )?;
+        // We used some asset for the swap, so we have to subtract that from the amount.
+        let remaining_asset_amount = amount.saturating_sub(consumed);
```

Comment:
Could it be additive instead of subtractive? I think if the sender agrees on `create_destination` it understands better the mechanics of the transfer and might be willing to cover the ED, the recipient on the other hand might be a new user how doesn't understand why he's getting 99.99 instead of the 100 he agreed to with the sender.

Reply:
Yeah, could be an option. That's getting more into the implementation detail. I think it gets to the core question of this RFC: Should AH directly add in a path to create accounts via Asset Conversion, or rely on people doing something like batching multiple calls together to handle the destination's existence? IMO, yes, with configuration options like "sender pays, receiver pays, fail if nonexistent".

Reply:
Yeah, this is now handled on the frontend by "guesstimating" the destination fee and adding it to the amount "sent"